### DISCUSSION TITLE
Memory allocation for Message

### DISCUSSION THREAD

**Discussion by jwinarske** [hyperium/tonic#1104] [Q&A]
I'm looking at a multi-language scenario.  Where memory is being shared between Rust and Dart.  It is possible to use a "zero copy buffer" and read/write from both sides.

Enter Tonic/Prost.

Is there a way to allocate the message buffers using a custom type such as a "zero copy buffer"?  The goal is to avoid any additional memory operations - move/alloc/copy.

Dart side would simply encode/decode protobuf messages using the "zero copy buffer".

**Comment by LucioFranco** [comment]
The `Message` trait accepts a `BufMut`/`Buf` from the `bytes` crate. You can use these generic abstractions to provide what ever buffer you want to encode/decode.

**Reply by jwinarske** [reply]
Is there a working example of this anywhere?

**Reply by LucioFranco** [reply]
Not that I know of but the abstractions are there so if you just map your buffers to that it should work.
